pallets/payment/src/lib.rs

```diff
@@ -415,43 +497,66 @@ pub mod pallet {
 					}
 					// Calculate incentive amount - this is to insentivise the user to release
 					// the funds once a transaction has been completed
-					let incentive_amount = T::IncentivePercentage::get().mul_floor(amount);
+					let incentive_amount = match charge_incentive_amount {
+						true => T::IncentivePercentage::get().mul_floor(amount),
+						false => 0u32.into(),
+					};
 
 					let mut new_payment = PaymentDetail {
 						asset,
 						amount,
 						incentive_amount,
-						state: PaymentState::Created,
+						state: payment_state,
 						resolver_account: T::DisputeResolver::get_origin(),
 						fee_detail: None,
 						remark,
 					};
 
 					// Calculate fee amount - this will be implemented based on the custom
-					// implementation of the marketplace
+					// implementation of the fee provider
 					let (fee_recipient, fee_percent) =
 						T::FeeHandler::apply_fees(&from, &recipient, &new_payment);
 					let fee_amount = fee_percent.mul_floor(amount);
 					new_payment.fee_detail = Some((fee_recipient, fee_amount));
 
-					// reserve the incentive amount from the payment creator
-					T::Asset::reserve(asset, &from, incentive_amount + fee_amount)?;
-					// transfer amount to recipient
-					T::Asset::transfer(asset, &from, &recipient, amount)?;
-					// reserved the amount in the recipient account
-					T::Asset::reserve(asset, &recipient, amount)?;
-
-					*maybe_payment = Some(new_payment);
+					*maybe_payment = Some(new_payment.clone());
 
-					Self::deposit_event(Event::PaymentCreated { from, asset, amount });
-					Ok(())
+					Ok(new_payment)
 				},
 			)
 		}
 
+		/// The function will reserve the fees+transfer amount from the `from` account. After reserving
+		/// the payment.amount will be transferred to the recipient but will stay in Reserve state.
+		#[require_transactional]
+		fn reserve_payment_amount(
+			from: &T::AccountId,
+			to: &T::AccountId,
+			payment: PaymentDetail<T>,
+		) -> DispatchResult {
+			let fee_amount = match payment.fee_detail {
```

Comment:
wondering if `map` and `unwrap_or_default` would be shorter ... at least they are sexier 😂